src/client/pfs_new_storage.go

```diff
@@ -34,16 +35,99 @@ func (c APIClient) PutTar(repo, commit string, r io.Reader) (retErr error) {
 	return err
 }
 
-// GetTar gets a tar stream out of PFS.
+// GetTar gets a tar stream out of PFS that contains files at the repo and commit that match the path.
 // Note: this should only be used for testing the new storage layer.
-func (c APIClient) GetTar(repo, commit, path string, w io.Writer) error {
+func (c APIClient) GetTar(repo, commit, path string) (io.Reader, error) {
 	req := &pfs.GetTarRequest{
 		File: NewFile(repo, commit, path),
 	}
-	gtc, err := c.PfsAPIClient.GetTar(c.Ctx(), req)
+	client, err := c.PfsAPIClient.GetTar(c.Ctx(), req)
 	if err != nil {
-		return grpcutil.ScrubGRPC(err)
+		return nil, grpcutil.ScrubGRPC(err)
+	}
+	return grpcutil.NewStreamingBytesReader(client, nil), nil
+}
+
+// GetTarConditional functions similarly to GetTar with the key difference being that each file's content can be conditionally downloaded.
+// GetTarConditional takes a callback that will be called for each file that matched the path.
+// The callback will receive the file information for the file and a reader that will lazily download a tar stream that contains the file.
+// Note: this should only be used for testing the new storage layer.
+func (c APIClient) GetTarConditional(repoName string, commitID string, path string, f func(fileInfo *pfs.FileInfoNewStorage, r io.Reader) error) error {
+	client, err := c.PfsAPIClient.GetTarConditional(c.Ctx())
+	if err != nil {
+		return err
+	}
+	if err := client.Send(&pfs.GetTarConditionalRequest{File: NewFile(repoName, commitID, path)}); err != nil {
+		return err
+	}
+	for {
+		resp, err := client.Recv()
+		if err != nil {
+			if err == io.EOF {
+				return nil
+			}
+			return err
+		}
+		r := &getTarConditionalReader{
+			client: client,
+			first:  true,
+		}
+		if err := f(resp.FileInfo, r); err != nil {
+			return err
+		}
+		if r.first {
+			if err := client.Send(&pfs.GetTarConditionalRequest{Skip: true}); err != nil {
+				return err
+			}
+			continue
+		}
+		if err := r.drain(); err != nil {
+			return err
+		}
+	}
+	return nil
+
+}
+
+type getTarConditionalReader struct {
+	client      pfs.API_GetTarConditionalClient
+	r           *bytes.Reader
+	first, done bool
+}
+
+func (r *getTarConditionalReader) Read(data []byte) (int, error) {
+	if r.first {
+		if err := r.client.Send(&pfs.GetTarConditionalRequest{Skip: false}); err != nil {
+			return 0, err
+		}
+		r.first = false
+	}
+	if r.r == nil || r.r.Len() == 0 {
+		if err := r.nextResponse(); err != nil {
+			return 0, err
+		}
+	}
+	return r.r.Read(data)
+}
+
+func (r *getTarConditionalReader) nextResponse() error {
```

Comment:
It seems like this should check if `r.done == true` and if it is return `io.EOF` immediately. I think in general you're supposed to be able to repeatedly read from a `Reader` that's hit `io.EOF` and keep getting `io.EOF`s from it.

Reply:
Still think this should be done.

Reply:
It is in the PR now.